### ISSUE TITLE
Possible Error in Code

### ISSUE THREAD

**Issue by kensun619** [thtrieu/darkflow#646]
In net/yolo/data.py and net/yolov2/data.py, when you calculate placeholder values, you will overwrite previous boxes. 
Specially
for obj in allobj:
        probs[obj[5], :] = [0.] * C
        probs[obj[5], labels.index(obj[0])] = 1.
        proid[obj[5], :] = [1] * C 
        coord[obj[5], :, :] = [obj[1:5]] * B
        prear[obj[5],0] = obj[1] - obj[3]**2 * .5 * S # xleft
        prear[obj[5],1] = obj[2] - obj[4]**2 * .5 * S # yup
        prear[obj[5],2] = obj[1] + obj[3]**2 * .5 * S # xright
        prear[obj[5],3] = obj[2] + obj[4]**2 * .5 * S # ybot
        confs[obj[5], :] = [1.] * B # If more two annoations fall in the same grid, the latter ones will overwrite previous ones?
Is this understanding correct?

**Comment by tianyu-tristan** [comment]
@kensun619 That line of code is just preparing confs as shape of (H*W, B). If the same grid cell has two annotations (i.e. true boxes), you're just overwriting that cell of [1.] * B by [1.] * B, which looks fine by me... But if the question is, how the model will perform if one grid cell has two true boxes (especially of two different classes), worth thinking...

**Comment by tianyu-tristan** [comment]
@kensun619 Oh i see, sorry that I kept thinking about PASCAL and COCO. I'm curious why your training dataset has two object in the same cell? Is it meant to be like this (e.g. two different object) or it can be optimized (e.g. if those two objects are in fact the same, run NMS to merge them)? In the case that, two different object falls under the same grid cell, there're still two cases: (1) those two objects match the same anchor box, or (2) those two objects match two different anchor box. If (1), the mechanics will be the same as you stated, latter overwrite the former; if (2), it is still fine because the regression target for confidence in loss function works per grid cell per anchor box predictor. But in general, I would recommend you take a look at YOLOv1 paper chapter 2.4 about limitation of YOLO, probably your application is not a best fit for YOLO.

**Comment by kensun619** [comment]
Remember the reason why they have a parameter for number of boxes is to deal with the situation. Since they down sample their images, two different cars, or one car one pedestrian will fall in one grid. These two annotation have complete different anchor box. If you overwrite them, the loss for overwritten annotation will be wrong.

**Comment by tianyu-tristan** [comment]
@kensun619 I don't think I get your point. Down sample images makes e.g. 1280x1280 becomes 416x416, the proportion is kept. Dividing your image into 13x13 grid cells, if you have a car and a person in 1280x1280, you should have the same car and same person in the same position (in terms of % of (width, height) in the "down sampled" image. For me, the chances that the center of the car and center of the person falls in the exact same grid cell out of 13x13 is not that likely, unless you have a very tight crowd of car/person. But even if that happens, the best YOLO can do for this scenario is what I said earlier, based on my understanding

**Comment by kensun619** [comment]
Check out this image. The further away the cars, the more likely it will fall in the same grid.

**Comment by tianyu-tristan** [comment]
@kensun619 in that case, it's small objects (e.g. flocks of birds) which YOLO is not dealing well with. I'm not arguing those situation should not happen, I'm just saying probably YOLO is not the best method to solve your problem. But I don't see a problem here in the code, to my best of knowledge

**Comment by kensun619** [comment]
OK. I got your point, but what is the reason to put the parameter that saying yolo is predicting 2 boxes (in yolo) or 3 boxes for each scale (in yolo v3).

**Comment by tianyu-tristan** [comment]
@kensun619 I didn't look at v3 yet, the 2 boxes in yolo is the number of anchor boxes B, whereby the output from yolo model is S x S x (B * 5 + C) stated in the paper, despite the code implementation here is giving S x S x B * (5 + C)

**Comment by kensun619** [comment]
Yes, I agree with you. But I want to confirm that yolo has the capability to deal with multiple boxes in one grid. This tensorflow implementation is not used to report the performance in any yolo paper. Is this how they implement in their C++ code?

**Comment by tianyu-tristan** [comment]
@kensun619 I don't know. For understanding, probably it's better to look at c code directly. I didn't do that:)

**Comment by kensun619** [comment]
I went through the c++ source code. I agree with you in Yolo v1 and Yolo v2, in each grid, one box can be predicted for one class. However in yolo v3, I think they changed how they store the ground truth. So for each predicted box, I think they are searching the ground truth box with highest IOU in the whole image instead of only considering boxes in the grid. 

PS: Are you interested in implementing other loss metric such as precision and recall? Also, how about adding validation?
